proposals/sns-ip-3.md

```diff
@@ -0,0 +1,124 @@
++                                                      RoA verification Id                                                      +
+                                                               ...                                                               
++                                                          (G*8 bytes)                                                          +
+|                                                                                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                                                                                                                               |
++                                                        Record Content                                                         +
+                                                               ...                                                               
++                                                          (C*8 bytes)                                                          +
+|                                                                                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ 
+```
+
+### Certification smart contract interface
+
+All supported record types will have at most two associated validation steps (one for each motivating concern).
+The first instruction will publish the pending record contents with `None` validation types.
+Then each validation step will be performed in turn.
+Any edit to the record contents will switch the validations to `None`.
+In theory, the entire validation can be performed in one transaction containing three instructions (or more in the case of Ethereum validation).
+
+One useful side-effect of using runtime Solana signing validation is that smart contract PDAs can be certification authorities, in which case a CPI call to `ValidateSolanaSignature` will perform the validation.
+
+```rust
+
+#[repr(u32)]
+pub enum Validation {
+    None,
+    SolanaSignature,
+    EthereumSignature,
+    ..
+}
+
+struct RecordBufferHeader {
+    staleness_validation_type: Validation,
+    staleness_validation_id: [u8;32],
+    right_of_association_validation: Validation
+}
+
+pub enum Instruction {
+    AllocateRecord {length: usize },
+    AllocateAndPostRecord { contents: Vec<u8> },
+    EditRecord { contents: Vec<u8>, offset: usize },
+    // This instruction will check that the proper authority signed the transaction
+    ValidateSolanaSignature,
+    // This instruction will use the ed25513 validation flow natively supported by Solana
+    // https://docs.solana.com/developing/runtime-facilities/programs#ed25519-program
```

Comment:
Small typo, should be https://docs.solana.com/developing/runtime-facilities/programs#secp256k1-program and `Secp256k1` instead of `ed25513`